cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -705,45 +702,21 @@ private ApplicationWithPrograms deployApp(NamespaceId namespaceId, @Nullable Str
    * @throws Exception
    */
   private void deleteApp(final ApplicationId appId, ApplicationSpecification spec) throws Exception {
-    Id.Application idApplication = appId.toId();
-
     //Delete the schedules
     scheduler.deleteSchedules(appId);
+    for (WorkflowSpecification workflowSpec : spec.getWorkflows().values()) {
+      scheduler.modifySchedulesTriggeredByDeletedProgram(appId.workflow(workflowSpec.getName()));
```

Comment:
who starts the transaction when it's called like this? Does this work?

Reply:
what happens when the same schedule is affected by multiple apps that are being deleted at the same time?

Reply:
the scheduler starts a transaction internally. There will be transaction conflict and the first successful retry will delete the schedule. Other transactions will not see the deleted schedules afterwards.